composable-tasks/src/compose_tasks.rs

```diff
@@ -161,6 +172,157 @@ pub trait TaskCall:
+            ERROR_CANNOT_SWAP_EGLD
+        );
+        require!(
+            args.len() >= SMART_SWAP_MIN_ARGS_LEN,
+            ERROR_SMART_SWAP_TWO_ARGUMENTS
+        );
+
+        let payment_in = payment_for_current_task.unwrap_esdt();
+        let mut amount_out = BigUint::zero();
+        let token_out = self.get_token_out_from_smart_swap_args(args.clone());
+
+        let mut args_iter = args.into_iter();
+
+        // First argument: number of swap operations
+        let num_operations_buf = args_iter
+            .next()
+            .unwrap_or_else(|| sc_panic!(ERROR_MISSING_NUMBER_OPS));
+        let num_operations = num_operations_buf
+            .parse_as_u64()
+            .unwrap_or_else(|| sc_panic!(ERROR_INVALID_NUMBER_OPS));
+
+        let mut final_payments = ManagedVec::new();
+        let mut acc_ammount_in = BigUint::zero();
+
+        // Avout out of gas issues
+        require!(
+            num_operations < SMART_SWAP_MAX_OPERATIONS,
+            "Provided too many operations for smart swap"
+        );
+        // Parse each operation
+        for _ in 0..num_operations {
+            // Parse: partial_amount_in, num_swap_ops, then swap operations
+            let partial_amount_in = BigUint::from(
+                args_iter
+                    .next()
+                    .unwrap_or_else(|| sc_panic!(ERROR_MISSING_AMOUNT_IN)),
+            );
+
+            acc_ammount_in += &partial_amount_in;
+
+            // Build swap arguments for this operation
+            let operation_swap_args = self.compose_smart_swap_operation_swap_args(&mut args_iter);
+
+            let operation_payment = EsdtTokenPayment::new(
+                payment_in.token_identifier.clone(),
+                payment_in.token_nonce,
+                partial_amount_in,
+            );
+
+            let mut operation_result = self.multi_pair_swap(operation_payment, operation_swap_args);
+            // Remove last payment; only residuals added
+            let partial_payment_out = operation_result.take(operation_result.len() - 1);
+            amount_out += partial_payment_out.amount;
+            final_payments.append_vec(operation_result);
+        }
+
+        require!(
+            acc_ammount_in <= payment_in.amount,
+            ERROR_ACC_AMOUNT_EXCEEDS_PAYMENT_IN
+        );
+
+        // Handle remaining amount if total percentage < 100%
+        if acc_ammount_in < payment_in.amount {
+            let remaining_amount = payment_in.amount - acc_ammount_in;
+
+            final_payments.push(EsdtTokenPayment::new(
+                payment_in.token_identifier.clone(),
+                payment_in.token_nonce,
+                remaining_amount,
+            ));
+        }
+
+        // Return the others to payments_to_return
+        payments_to_return.append_vec(final_payments);
+
+        let fee_percentage = self.smart_swap_fee_percentage().get();
+        let fee_taken = amount_out.clone() * fee_percentage;
```

Comment:
If you update the smart_swap_fees key to a &EgldOrEsdtTokenIdentifier, you can rewrite this as so (in order to avoid multiple clones:

```
        let fee_taken = &amount_out * fee_percentage;
        self.smart_swap_fees(&token_out)
            .update(|total_fees| *total_fees += &fee_taken);
        let remaining_amount_after_fee = amount_out - fee_taken;

        EgldOrEsdtTokenPayment::new(token_out, 0, remaining_amount_after_fee)
```

Reply:
Kept it as `TokenIdentifier`, but  managed to get rid of a clone.

Reply:
Fixed.